element/serialize.js

```diff
@@ -393,52 +454,83 @@ export function renderElement( element, context = {} ) {
 /**
  * Serializes a native component type to string.
  *
- * @param {?string} type    Native component type to serialize, or null if
- *                          rendering as fragment of children content.
- * @param {Object}  props   Props object.
- * @param {?Object} context Context object.
+ * @param {string}  type        Native component type to serialize.
+ * @param {Object}  props       Props object.
+ * @param {?Object} context     Context object.
+ * @param {?number} indentLevel In recursion, level at which to indent.
  *
  * @return {string} Serialized element.
  */
-export function renderNativeComponent( type, props, context = {} ) {
-	let content = '';
+export function renderNativeComponent( type, props, context, indentLevel ) {
+	let childrenContent = '';
+
+	let childrenIndentLevel = indentLevel;
+	if ( ! isInlineTag ) {
+		childrenIndentLevel++;
+	}
+
 	if ( type === 'textarea' && props.hasOwnProperty( 'value' ) ) {
 		// Textarea children can be assigned as value prop. If it is, render in
 		// place of children. Ensure to omit so it is not assigned as attribute
 		// as well.
-		content = renderChildren( [ props.value ], context );
+		childrenContent = renderChildren( [ props.value ], context, childrenIndentLevel );
 		props = omit( props, 'value' );
 	} else if ( props.dangerouslySetInnerHTML &&
 			typeof props.dangerouslySetInnerHTML.__html === 'string' ) {
 		// Dangerous content is left unescaped.
-		content = props.dangerouslySetInnerHTML.__html;
+		childrenContent = props.dangerouslySetInnerHTML.__html;
 	} else if ( typeof props.children !== 'undefined' ) {
-		content = renderChildren( castArray( props.children ), context );
+		childrenContent = renderChildren( castArray( props.children ), context, childrenIndentLevel );
 	}
 
 	if ( ! type ) {
-		return content;
+		return childrenContent;
+	}
+
+	let content = '';
+
+	// Place non-inline tag on own line with indentation.
+	const isInlineTag = INLINE_TAGS.has( type );
+	if ( ! isInlineTag && indentLevel > 0 ) {
+		content += '\n' + repeat( '\t', indentLevel );
 	}
 
 	const attributes = renderAttributes( props );
 
-	if ( SELF_CLOSING_TAGS.has( type ) ) {
-		return '<' + type + attributes + '/>';
+	if ( type ) {
+		content += '<' + type + attributes;
+
+		if ( SELF_CLOSING_TAGS.has( type ) ) {
+			return content + '/>';
```

Comment:
Should we add a space for self closing tags?

Reply:
>Should we add a space for self closing tags?

Heh, I went back and forth on this a few times. I don't feel particularly strongly about it, and ultimately weighed consistency of this "beauty" before deciding to remove it: Why is there a space after the last attribute when a tag is self-closing, but not when it has a separate explicit closing tag? Why do we otherwise not include the space before slash in a closing tag, e.g. `< /figure>` ?

I'd agree that at least by anecdotal observation, the space before the self-closing slash is common for well-formatted HTML, though I couldn't personally come to a conclusion as to why this is objectively the case.

I'll see if I can add it without conflicts, at least so that we're consistent with the previous output, and if future debate is warranted, I'm open to discussing further.

Reply:
It doesn't matter to me. Just thought it would avoid some changes in the tests. Maybe we can also drop the `/` if it only needs to be HTML5 compatible. I guess not since not all themes will be HTML5.